lnwallet/rpcwallet/sign_coordinator.go

```diff
@@ -0,0 +1,1039 @@
+	// ErrRequestTimeout is the error that's returned if we time out while
+	// waiting for a response from the remote signer.
+	ErrRequestTimeout = errors.New(
+		"remote signer response timeout reached")
+
+	// ErrConnectTimeout is the error that's returned if we time out while
+	// waiting for the remote signer to connect.
+	ErrConnectTimeout = errors.New(
+		"timed out when waiting for remote signer to connect ")
+
+	// ErrMultipleConnections is the error that's returned if another
+	// remote signer attempts to connect while we already have one
+	// connected.
+	ErrMultipleConnections = errors.New(
+		"only one remote signer can be connected")
+
+	// ErrNotConnected is the error that's returned if the remote signer
+	// closes the stream or we encounter an error when receiving over the
+	// stream.
+	ErrNotConnected = errors.New("the remote signer is no longer connected")
+
+	// ErrUnexpectedResponse is the error that's returned if the response
+	// with the expected request ID from the remote signer is of an
+	// unexpected type.
+	ErrUnexpectedResponse = errors.New("unexpected response type")
+)
+
+// SignCoordinator is an implementation of the signrpc.SignerClient and the
+// walletrpc.WalletKitClient interfaces that passes on all requests to a remote
+// signer. It is used by the watch-only wallet to delegate any signing or ECDH
+// operations to a remote node over a
+// walletrpc.WalletKit_SignCoordinatorStreamsServer stream. The stream is set up
+// by the remote signer when it connects to the watch-only wallet, which should
+// execute the Run method.
+type SignCoordinator struct {
+	// stream is a bi-directional stream between us and the remote signer.
+	stream walletrpc.WalletKit_SignCoordinatorStreamsServer
+
+	// responses is a map of request IDs to response channels. This map
+	// should be populated with a response channel for each request that has
+	// been sent to the remote signer. The response channel should be
+	// inserted into the map before the request is sent.
+	// Any response received over the stream that does not have an
+	// associated response channel in this map is ignored.
+	// The response channel should be removed from the map when the response
+	// has been received and processed.
+	responses map[uint64]chan *walletrpc.SignCoordinatorResponse
+
+	// receiveErrChan is used to signal that the stream with the remote
+	// signer has errored, and we can no longer process responses.
+	receiveErrChan chan error
+
+	// doneReceiving is closed when either party terminates and signals to
+	// any pending requests that we'll no longer process the response for
+	// that request.
+	doneReceiving chan struct{}
+
+	// quit is closed when lnd is shutting down.
+	quit chan struct{}
+
+	// clientConnected is sent over when the remote signer connects.
+	clientConnected chan struct{}
+
+	// nextRequestID keeps track of the next request ID that should
+	// be used when sending a request to the remote signer.
+	nextRequestID uint64
```

Comment:
This could be an `atomic.Uint64`.

Reply:
Thanks! That made things quite a bit more clean :)